### PR TITLE
Fix almost all game related issues

### PR DESCRIPTION
**Pull request by damianatorrpm**
Requires https://github.com/swaywm/wlroots/pull/2264

* Adds minimizing support for chrome / steam csd
  and games that want to minimize on focus loss
* Adds view-self-focus-request for xwayland like gtk-shell has
* Adds view-hints-changed-signal (for urgency hint, possibly others)
* Adds view-system-bell signal (currently used in gtk-shell only)
* Fixes fullscreen games, the move behavior of fullscreen views,
   is to snap back to fullscreen after move. I would demo how 
  smooth and nice this works with multiple outputs but wf-recorder
  can only record one.
* Disallow resizing & snap off for fullscreen as this breaks 95 % of steam games,
   since they expect the size to be the resolution it is set to.

I have to say that in some occasions the switcher plugin crashes wayfire,
but I am unable to find the cause. The crash never happens if you 
reload the wobbly plugin at the beginning of your session,
maybe the load order is important here...

The fast switcher work perfectly fine.
It seems to be a 'floating point exception in switcher' and I found switcher
to also have other kind of strange behavior as setting views alpha to 0.0
when minimized and sometimes after restore it is not restored to 1.0,
also the other issue I demoed with the switcher (can't switch).

I have disabled this plugin and hope for a soon present plugin :100:

### CODE REVIEW

**Comment by damianatorrpm** [comment]
Confirming: the order is important - wobbly plugin must be before switcher plugin otherwise a rare issue might crash wayfire

File: src/api/wayfire/signal-definitions.hpp [hunk 1]
```diff
@@ -75,6 +75,22 @@ struct view_self_request_focus_signal : public _view_signal
 {
 };
 
+/**
+ * view-hints-changed signal is emitted on the view's output whenever the client indicates the
+ * views hints have changed (example urgency hint).
```
Comment (ammen99) [inline src/api/wayfire/signal-definitions.hpp:80]: What is the difference between hints-changed and system-bell, except that they come from different sources?

File: src/view/view-impl.cpp [hunk 2]
```diff
@@ -227,7 +227,7 @@ wlr_surface *wf::wlr_view_t::get_keyboard_focus_surface()
 
 bool wf::wlr_view_t::should_be_decorated()
 {
-    return role == wf::VIEW_ROLE_TOPLEVEL && !has_client_decoration;
+    return role == wf::VIEW_ROLE_TOPLEVEL && !has_client_decoration && !fullscreen;
```
Comment (ammen99) [inline src/view/view-impl.cpp:230]: the decoration plugin is responsible for this, dont' put it here. If you see some bug related to this, let's look into how to fix it in the decoration plugin.

File: src/view/xwayland.cpp [hunk 3]
```diff
@@ -305,9 +306,24 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
...
+                this->get_output()->emit_signal("view-self-request-focus", &data);
+            }
+        });
+        on_request_minimize.set_callback([&] (void*) {
+            minimize_request(xw->minimized);
+        });
         on_request_maximize.set_callback([&] (void*) {
-            tile_request((xw->maximized_horz && xw->maximized_vert) ?
-                wf::TILED_EDGES_ALL : 0);
+            if (xw->maximized_horz && xw->maximized_vert)
```
Comment (ammen99) [inline src/view/xwayland.cpp:322]: wait why the change? It doesn't do anything ...

Comment (ammen99) [reply src/view/xwayland.cpp:322]: I mean it is the same as before

File: src/view/xwayland.cpp [hunk 4]
```diff
@@ -327,11 +343,29 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
             update_decorated();
         });
 
+        on_set_hints.set_callback([&] (void*) {
+            /*
+             * By storing urgency data instead of a signal
+             * it can be manually "acknowledged" by removing the data
+            */
+            if (xw->hints_urgency && !this->has_data("view-demands-attention"))
+                this->store_data(std::unique_ptr<wf::custom_data_t>(), "view-demands-attention");
```
Comment (ammen99) [inline src/view/xwayland.cpp:352]: why is this useful? I think that emitting the signal is enough. Plugins can store this data themselves if they want/need to.

File: src/view/xwayland.cpp [hunk 5]
```diff
@@ -400,7 +438,10 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
         }
 
         if (xw->fullscreen)
-            fullscreen_request(get_output(), true);
+        {
+                this->view_impl->last_windowed_geometry = get_output_geometry();
```
Comment (ammen99) [inline src/view/xwayland.cpp:442]: 1) output geometry is the wrong thing to take, you should pick wm_geometry
2) what bug(s) does this fix? This should already be handled in fullscreen_request

File: plugins/single_plugins/move-snap-helper.hpp [hunk 6]
```diff
@@ -106,16 +106,13 @@ class move_snap_helper_t : public wf::custom_data_t
   protected:
     virtual bool should_enable_snap_off() const
     {
-        return enable_snap_off &&
```
Comment (ammen99) [inline plugins/single_plugins/move-snap-helper.hpp:109]: I don't see how this change is useful. If snap-off is enabled, this means the view stays in place until a certain threshold is achieved - looks like exactly what we want for fullscreen?

File: src/api/wayfire/signal-definitions.hpp (hunk 1 above)
Comment (damianatorrpm) [reply src/api/wayfire/signal-definitions.hpp:80]: well, if you have gnome terminal and test it on gnome, you will see that if your cursor is already most-left at the terminal and press left you will hear woof if that is your alert sound.
hints changed is really like IM message received

File: src/view/xwayland.cpp (hunk 3 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp:322]: You looked at the newer commits which almost all say test,
because of the switcher plugin :)

There is one other issue I have been looking into:
https://github.com/ValveSoftware/steam-for-linux/issues/3792

this change was basically a result though that these are different properties and possibly at some point this should be changed further to actually make partial maximize recognized (i.e horz / vert)

File: src/view/xwayland.cpp (hunk 4 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp:352]: kinda unsure since hints_change may represent !xw->hints_urgency at some point, than what? view-no-longer-demands-attention?

File: src/view/xwayland.cpp (hunk 5 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp:442]: none really as mentioned above a desperate try to see why swticher crashes.

File: plugins/single_plugins/move-snap-helper.hpp (hunk 6 above)
Comment (damianatorrpm) [reply plugins/single_plugins/move-snap-helper.hpp:109]: the important part here is that it removes the condition || fullscreen.
`If snap-off is enabled, this means the view stays in place until a certain threshold is achieved - looks like exactly what we want for fullscreen?`
Not sure what you mean but a definitive no.
With the patch you can freely move fullscreen views between outputs etc.
But snap-off resizes and un-fullscreens a fullscreen view because it thinks that right, 
which is not, hence snap-off is disabled for fullscreen.
It's just how applications work, hence the long comment, 
when you take games they decide (via option) if fullscreen or not, not the compositor.

File: src/view/view-impl.cpp (hunk 2 above)
Comment (damianatorrpm) [reply src/view/view-impl.cpp:230]: fair enough, will remove and we'll fix whatever happens

File: plugins/single_plugins/move.cpp [hunk 7]
```diff
@@ -273,6 +292,17 @@ class wayfire_move : public wf::plugin_interface_t
             workspace_may_changed.old_viewport_invalid = false;
             output->emit_signal("view-change-viewport", &workspace_may_changed);
 
+            /*
+             * Restore the fullscreen state so the window is
+             * not in a weird position afterwards, fullscreen
+             * are by spec the geometry of the output (or span outputs)
+             */
+            if (this->view->fullscreen && this->view->activated
```
[the remark this answers is not in the document]
Comment (damianatorrpm) [reply plugins/single_plugins/move.cpp:300]: these are actually too many checks this->view->fullscreen was enough but,
remember the minimize bug? clicking on the surface that is being minimized?
well this is a result of finding out why dragging a different view, during the minimize action of a game that has "minimize on focus loss" actually is restored if a different view is dragged during it's minimize animation.

File: src/view/xwayland.cpp (hunk 3 above)
Comment (ammen99) [reply src/view/xwayland.cpp:322]: Just look at the code before and now, you have not changed ANYTHING.

Comment (ammen99) [reply src/view/xwayland.cpp:322]: Well, except that before it was using ternary operator and now it uses if/else ..

File: src/view/xwayland.cpp (hunk 4 above)
Comment (ammen99) [reply src/view/xwayland.cpp:352]: put a flag in the signal?

File: src/view/xwayland.cpp (hunk 5 above)
Comment (ammen99) [reply src/view/xwayland.cpp:442]: alright, then, since this doesn't change anything, revert the change, most probably the bug is in switcher itself.

File: src/view/xwayland.cpp (hunk 4 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp:352]: you're right, my usual thinking is more toward the goal, whether my mind is currently located in plugin or not. The (theoretical, sure let's not get ahead of ourselves) issue I have is this though,
this requires all plugins that use this to use the same semantics for the data i.e the data name,
which hasn't been 'defined' anywhere. I understand this may not seem real at this point but the virtual barrier this detail puts, well I can't unsee it...

Comment (damianatorrpm) [reply src/view/xwayland.cpp:352]: Nevertheless of course I will change it too a flag in the signal

File: plugins/single_plugins/move-snap-helper.hpp (hunk 6 above)
Comment (ammen99) [reply plugins/single_plugins/move-snap-helper.hpp:109]: Snap-off is not responsible for unfullscreening and resizing the view! Snap-off is the thing which holds the view in place, so that you don't accidentally move it. 

However, I can see what you mean. Wouldn't it be better to just not unfullscreen the view in the `snap_off()` function? This way the snap-off-threshold is still respected (as it should) but the view is not unfullscreened.

Comment (damianatorrpm) [reply plugins/single_plugins/move-snap-helper.hpp:109]: I can see what you mean. Practically it's that you can't start a move operation by the decoration,
for fullscreen views since the have no decoration (maybe with some rare exceptions though) 
and one has to use the shortcut. Nevertheless I absolutely agree that it should respect the snap-off-threshold.

File: src/view/xwayland.cpp (hunk 4 above)
Comment (ammen99) [reply src/view/xwayland.cpp:352]: But each plugin that cares about this data will receive the signal, right? So all of them will know that the view has or doesn't have the urgent hint.

File: src/view/xwayland.cpp (hunk 3 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp:322]: If you implement at one point vert/horz maximization in this function it will have to change from the ternary operator proably. I wanted to implement this but wlroots has no signal for vert/horz tile changes yet. I can revert this.

File: plugins/single_plugins/move.cpp [hunk 8]
```diff
@@ -210,11 +210,30 @@ class wayfire_move : public wf::plugin_interface_t
...
+             * It is possible to make the below clause an option,
+             * but not having that option enabled will break any
+             * game that is fullscreen or set's their own size,
+             * as tiling will resize them.
+             *
+             * This way fullscreened view can be moved from one
+             * output to another and be restored to fullscreen
+             * on a different output (hopefully with the same resolotion)
+             */
+            if (enable_snap && !view->fullscreen)
```
[the remark this answers is not in the document]
Comment (damianatorrpm) [reply plugins/single_plugins/move.cpp:236]: Will see now :)

File: src/view/xwayland.cpp (hunk 3 above)
Comment (ammen99) [reply src/view/xwayland.cpp:322]: I'd keep it as it was before, until the moment we implement separate maximization.

File: src/api/wayfire/signal-definitions.hpp (hunk 1 above)
Comment (ammen99) [reply src/api/wayfire/signal-definitions.hpp:80]: Maybe a stupid question, but is this different from self-focus-request? My understanding was that the two are very similar.

Comment (damianatorrpm) [reply src/api/wayfire/signal-definitions.hpp:80]: Yes it's different :) 
The hint is not necessarily a request for focus.
It can be interpreted by a notify-send, a blinking taskbar icon, a sound,
whatever.

Comment (damianatorrpm) [reply src/api/wayfire/signal-definitions.hpp:80]: Also I kept it as view-hints-changed not view-urgency or something as there maybe other kind of hints

File: src/view/xwayland.cpp (hunk 3 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp]: Done

File: src/output/gtk-shell.cpp [hunk 9]
```diff
@@ -147,7 +147,18 @@ static void handle_gtk_shell_set_startup_id(wl_client *client, wl_resource *reso
  *  A view could use this to invoke the system bell, be it aural, visual or none at all.
  *  Not implemented.
  */
-static void handle_gtk_shell_system_bell(wl_client *client, wl_resource *resource, wl_resource *surface) {}
+static void handle_gtk_shell_system_bell(wl_client *client, wl_resource *resource, wl_resource *surface)
+{
+    auto wl_surface = static_cast<wl_resource*> (wl_resource_get_user_data(surface));
```
Comment (kode54) [inline src/output/gtk-shell.cpp:152]: This entire function needs a null pointer guard on the `surface` parameter, and should somehow handle a case for firing a bell signal to the shell itself, or something else. I fixed a crash simply adding a null guard to my local copy, but that doesn't send a bell signal anywhere in that case.

Comment (ammen99) [reply src/output/gtk-shell.cpp:152]: seeing this, maybe it makes more sense to just emit the signal on core?

**Comment by ammen99** [comment]
You said you'll send separate PRs for these changes? Maybe we can close this one.
